File: SSUMobile/Modules/About/Views/SSUAboutViewController.m
//
//  SSUAboutViewController.m
//  SSUMobile
//
//  Created by Eric Amorde on 10/20/14.
//  Copyright (c) 2014 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUAboutViewController.h"
#import "SSUWebViewController.h"
#import "SSULDAPCredentials.h"
#import <SDWebImage/SDImageCache.h>
#import "MBProgressHUD.h"

static NSString * LEGAL_CELL = @"Legal";
static NSString * LDAP_CELL = @"LDAP";
static NSString * CACHE_CELL = @"Cache";

@interface SSUAboutViewController ()

@property (nonatomic, weak) IBOutlet UILabel * versionLabel;
@property (nonatomic, weak) IBOutlet UILabel * imageCacheLabel;
@property (nonatomic, weak) IBOutlet UILabel * ldapLabel;

@end

@implementation SSUAboutViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.estimatedRowHeight = UITableViewAutomaticDimension;
    
    self.versionLabel.text = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"];
    [self updateCacheDisplay];
    [self updateLDAPDisplay];
}

- (void) updateCacheDisplay {
    [[SDImageCache sharedImageCache] calculateSizeWithCompletionBlock:^(NSUInteger fileCount, NSUInteger totalSize) {
        double MB = totalSize / 1024.0 / 1024.0; // bytes --> MB
        self.imageCacheLabel.text = [NSString stringWithFormat:@"%0.2fMB / %lu images",MB,(unsigned long)fileCount];
    }];
}

- (void) updateLDAPDisplay {
    self.ldapLabel.enabled = [[SSULDAPCredentials sharedInstance] hasCredentials];
}

#pragma mark - Table View

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell * cell = [self tableView:tableView cellForRowAtIndexPath:indexPath];
    if ([cell.reuseIdentifier isEqualToString:LEGAL_CELL]) {
        SSUWebViewController * webView = [SSUWebViewController webViewController];
        webView.urlToLoad = [[NSBundle mainBundle] pathForResource:@"licenses" ofType:@"txt"];
        
        [self.navigationController pushViewController:webView animated:YES];
    }
    else if ([cell.reuseIdentifier isEqualToString:LDAP_CELL]) {
        if ([[SSULDAPCredentials sharedInstance] hasCredentials]) {
            [[SSULDAPCredentials sharedInstance] clearCredentials];
            [self updateLDAPDisplay];
            MBProgressHUD * hud = [MBProgressHUD showHUDAddedTo:self.tableView animated:YES];
            hud.labelText = @"Logged Out";
            hud.mode = MBProgressHUDModeText;
            [hud hide:YES afterDelay:1.0];
        }
    }
    else if ([cell.reuseIdentifier isEqualToString:CACHE_CELL]) {
        [[SDImageCache sharedImageCache] clearDiskOnCompletion:^{
            [self updateCacheDisplay];
        }];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}


@end

File: SSUMobile/Modules/About/Views/SSUAboutViewController.swift
//
//  SSUAboutViewController.swift
//  SSUMobile
//
//  Created by Eric Amorde on 3/25/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

class SSUAboutViewController: UITableViewController {
    
    struct ReuseIdentifier {
        static let legal = "Legal"
        static let ldap = "LDAP"
        static let cache = "Cache"
        static let feedback = "Feedback"
    }
    
    private let cacheFormatter: ByteCountFormatter = {
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = .useMB
        formatter.countStyle = .file
        formatter.allowsNonnumericFormatting = false
        return formatter
    }()
    
    @IBOutlet var versionLabel: UILabel!
    @IBOutlet var imageCacheLabel: UILabel!
    @IBOutlet var ldapLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.estimatedRowHeight = UITableViewAutomaticDimension
        
        versionLabel.text = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        updateCacheDisplay()
        updateLDAPDisplay()
    }
    
    private func updateCacheDisplay() {
        ImageCache.default.calculateDiskCacheSize { (bytes) in
            let mb = self.cacheFormatter.string(fromByteCount: Int64(bytes))
            self.imageCacheLabel.text = "Cached images: \(mb)"
        }
    }
    
    private func updateLDAPDisplay() {
        ldapLabel.isEnabled = SSULDAPCredentials.sharedInstance().hasCredentials
    }
    
    // MARK: UITableView
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        guard let cell = tableView.cellForRow(at: indexPath), let identifier = cell.reuseIdentifier else {
            SSULogging.logError("Unable to retrieve selected cell in AboutViewController")
            return
        }
        switch identifier {
        case ReuseIdentifier.legal:
            if let fileURL = Bundle.main.url(forResource: "licenses", withExtension: "txt") {
                let webVc = SSUWebViewController()
                webVc.urlToLoad = fileURL
                self.navigationController?.pushViewController(webVc, animated: true)
            }
        case ReuseIdentifier.ldap:
            if SSULDAPCredentials.sharedInstance().hasCredentials {
                SSULDAPCredentials.sharedInstance().clear()
                updateLDAPDisplay()
                let hud = MBProgressHUD.showAdded(to: tableView, animated: true)
                hud.label.text = "Logged Out"
                hud.mode = .text
                hud.hide(animated: true, afterDelay: 1.0)
            }
        case ReuseIdentifier.cache:
            let hud = MBProgressHUD.showAdded(to: tableView, animated: true)
            hud.label.text = "Clearing cache..."
            hud.mode = .annularDeterminate
            ImageCache.default.clearDiskCache(completion: { 
                hud.label.text = "Cache cleared"
                hud.mode = .text
                hud.hide(animated: true, afterDelay: 1.0)
                self.updateCacheDisplay()
            })
        case ReuseIdentifier.feedback:
            // Handled by storyboard
            break
        default:
            SSULogging.logError("Unrecognized cell identifier \(identifier)")
        }
    }
    
}
